flecsi/data/local/policy.hh

```diff
@@ -0,0 +1,556 @@
+    // vertex, edge, cell). We thus need to use the p.get_storage() to get the
+    // metadata, not the region.get_storage() which gives the real data. This
+    // works the same way as how one partition stores the number of elements in
+    // a particular 'shard' on a rank for another partition. In addition, we
+    // also need to copy Values from the partition and save them locally. User
+    // code might change it after this constructor returns. We can not use a
+    // copy assignment directly here since metadata is an util::span while
+    // ghost_ranges is a std::vector<>.
+    ghost_ranges = to_vector(p.template get_storage<Value>(fid));
+    // Get The largest value of `end index` in ghost_ranges (i.e. the upper
+    // bound). This tells how much memory needs to be allocated for ghost
+    // entities.
+    if(auto iter = std::max_element(ghost_ranges.begin(),
+         ghost_ranges.end(),
+         [](Value x, Value y) { return x.second < y.second; });
+       iter != ghost_ranges.end()) {
+      max_end = iter->second;
+    }
+  }
+
+private:
+  // This member function is only called by copy_engine.
+  friend struct copy_engine<Storage>;
+
+  template<typename T>
+  auto get_storage(field_id_t fid) const {
+    return r->template get_storage<T>(fid, max_end);
+  }
+
+  // We use a reference to region instead of pointer since it is not nullable
+  // nor needs to be redirected after initialization.
+  region_impl<Storage> * r;
+
+  // Locally cached metadata on ranges of ghost index.
+  std::vector<Value> ghost_ranges;
+  std::size_t max_end = 0;
+};
+
+template<typename Storage>
+struct points {
+  using Value = std::pair<std::size_t, std::size_t>; // (rank, index)
+  static Value make(std::size_t r, std::size_t i) {
+    return {r, i};
+  }
+
+  points(detail::prefixes<Storage> & p,
+    const intervals<Storage> &,
+    field_id_t,
+    completeness = incomplete)
+    : r(&p.get_base()) {}
+
+private:
+  // The region `r` contains field data of shared entities on this rank as
+  // source to be copied to remote peers. We make copy_engine a friend to allow
+  // direct access to the region.
+  friend struct copy_engine<Storage>;
+
+  region_impl<Storage> * r;
+};
+
+template<typename Storage>
+struct copy_engine {
+  // One copy engine for each entity type i.e. vertex, cell, edge.
+  copy_engine(const points<Storage> & points,
+    const intervals<Storage> & intervals,
+    field_id_t meta_fid /* for remote shared entities */)
+    : source(points), destination(intervals) {
+    // There is no information about the indices of local shared entities,
+    // ranks and indices of the destination of copy i.e. (local source
+    // index, {(remote dest rank, remote dest index)}). We need to do a shuffle
+    // operation to reconstruct this info from {(local ghost index, remote
+    // source rank, remote source index)}.
+    auto remote_sources =
+      destination.template get_storage<const typename points<Storage>::Value>(
```

Comment:
I got 

>/home/ollie/CLionProjects/flecsi/flecsi/data/local/policy.hh:439:55: error: expected a qualified >name after 'typename'
>      destination.template get_storage<const typename points<Storage>::Value>(

on Clang 13.0.0. 

Did we miss something here?

Reply:
Uhh, it compiled for me using gcc 10. I'll investigate.

Reply:
This is actually a very subtle name lookup issue, brought on here by having a function parameter and the nearby class template both named `points`.  I believe Clang is correct to reject it, but only as a result of my own recent paper on the subject.

Reply:
Why do we have to template parametrize everything? Since we are using `#include` and overload names for different backends, why not just let the HPX backend define it own version of `data_storage` and left everything done by `#include` the right header files in the right order.

Reply:
I have changed the code to use a backend-specific type instead of template arguments.

Reply:
I can compile it. However, I still see SEGFAULT.

> Thread 10 "index" received signal SIGSEGV, Segmentation fault.
> [Switching to Thread 0x7fffecff9640 (LWP 872521)]
> 0x0000000000597a41 in flecsi::data::mutator<(flecsi::data::layout)3, int, 6u>::raw_row::brk (this=0x7ffff4347d08) at /home/ollie/CLionProjects/flecsi/flecsi/data/accessor.hh:327
> 327	      return s.size() - o->del;
> (gdb) p *this
> $1 = {s = {p = 0x7fffcc056520, q = 0x7fffcc056520}, o = 0x0}
> (gdb) 

@opensdh 

Is the mutator thread-safe? Why is the "overflow" point being nullptr?

Reply:
What test fails, what command line arguments? What platform? Which compiler?

Reply:
The "index" unit test. "ctest -V -R index". Fedora 35, clang++ 13.0.0.

Reply:
Running under VMWare in MacOS.

Reply:
The "unstructured" unit test also fault at the same spot.

Reply:
```
==875057== Invalid read of size 8
==875057==    at 0x6A5671: flecsi::data::mutator<(flecsi::data::layout)3, unsigned int, 124u>::raw_row::brk() const (flecsi/data/accessor.hh:327)
==875057==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==875057== 
==875056== Thread 13:
==875056== Invalid read of size 8
==875056==    at 0x6A5671: flecsi::data::mutator<(flecsi::data::layout)3, unsigned int, 124u>::raw_row::brk() const (flecsi/data/accessor.hh:327)
==875056==  Address 0x0 is not stack'd, malloc'd or (recently) free'd

```

Reply:
The `Overflow` object is initialized by the backend, which installs it by calling `mutator<ragged,…>::buffer`.  All of this is handled by a [`param_buffers` object](https://github.com/flecsi/flecsi/blob/9779ed5d91d93e5e4025932b52d81f5cc82169de/flecsi/exec/mpi/policy.hh#L77).

Reply:
Just by comparing the implementation of `reduce_internal`, HPX backend does miss that finalize object (or it is been moved around and only exists on non-root).

@hkaiser 

Can you take a look at [it](https://github.com/flecsi/flecsi/blob/bba8fdc0de17349ef5ba671fa5d45d4c4d2313f9/flecsi/exec/hpx/policy.hh#L128)?